unittest/test_codegen.cpp

```diff
@@ -0,0 +1,479 @@
+
+  // Create a cost model per the running and terminal action model.
+  boost::shared_ptr<crocoddyl::CostModelSum> runningCostModel = boost::make_shared<crocoddyl::CostModelSum>(state);
+  boost::shared_ptr<crocoddyl::CostModelSum> terminalCostModel = boost::make_shared<crocoddyl::CostModelSum>(state);
+
+  // Then let's added the running and terminal cost functions
+  runningCostModel->addCost("gripperPose", goalTrackingCost, 1);
+  runningCostModel->addCost("xReg", xRegCost, 1e-4);
+  runningCostModel->addCost("uReg", uRegCost, 1e-4);
+  terminalCostModel->addCost("gripperPose", goalTrackingCost, 1);
+
+  // We define an actuation model
+  boost::shared_ptr<crocoddyl::ActuationModelFull> actuation =
+      boost::make_shared<crocoddyl::ActuationModelFull>(state);
+
+  // Next, we need to create an action model for running and terminal knots. The
+  // forward dynamics (computed using ABA) are implemented
+  // inside DifferentialActionModelFullyActuated.
+  boost::shared_ptr<crocoddyl::DifferentialActionModelFreeFwdDynamics> runningDAM =
+      boost::make_shared<crocoddyl::DifferentialActionModelFreeFwdDynamics>(state, actuation, runningCostModel);
+
+  boost::shared_ptr<crocoddyl::DifferentialActionModelFreeFwdDynamics> terminalDAM =
+      boost::make_shared<crocoddyl::DifferentialActionModelFreeFwdDynamics>(state, actuation, terminalCostModel);
+
+  // Eigen::VectorXd armature(state->get_nq());
+  // armature << 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.;
+  // runningDAM->set_armature(armature);
+  // terminalDAM->set_armature(armature);
+  boost::shared_ptr<crocoddyl::ActionModelAbstract> runningModel =
+      boost::make_shared<crocoddyl::IntegratedActionModelEuler>(runningDAM, 1e-3);
+  boost::shared_ptr<crocoddyl::ActionModelAbstract> terminalModel =
+      boost::make_shared<crocoddyl::IntegratedActionModelEuler>(terminalDAM, 1e-3);
+
+  std::vector<boost::shared_ptr<crocoddyl::ActionModelAbstract> > runningModels(N, runningModel);
+
+  boost::shared_ptr<crocoddyl::ShootingProblem> problem =
+      boost::make_shared<crocoddyl::ShootingProblem>(x0, runningModels, terminalModel);
+  std::vector<Eigen::VectorXd> xs(N + 1, x0);
+
+  std::vector<Eigen::VectorXd> us(N, Eigen::VectorXd::Zero(runningModel->get_nu()));
+  for (unsigned int i = 0; i < N; ++i) {
+    const boost::shared_ptr<crocoddyl::ActionModelAbstract>& model = problem->get_runningModels()[i];
+    const boost::shared_ptr<crocoddyl::ActionDataAbstract>& data = problem->get_runningDatas()[i];
+    model->quasiStatic(data, us[i], x0);
+  }
+
+  crocoddyl::SolverDDP ddp(problem);
+  ddp.setCandidate(xs, us, false);
+
+  /**************************ADScalar**********************/
+  /**************************ADScalar**********************/
+  /**************************ADScalar**********************/
+  typedef double Scalar;
+  typedef crocoddyl::MathBaseTpl<Scalar>::VectorXs VectorXs;
+
+  typedef CppAD::cg::CG<Scalar> CGScalar;
```

Comment:
I don't think you have to provide it in the main fwd.hpp.
I would rather suggest splitting the codegen from the main corpus. But this should be done later, I would suggest.

Reply:
I don't think splitting the codegen from main test might be useful, since the CGScalar is supposed to copy _exactly_ the definition of the action model that is being defined with Scalar.

In the future, there needs to be a `cast` function like in Eigen and Pinocchio, so that once an actionmodel is defined, it can simply be cast to a new type by .cast<NewScalar>(). However, it is kind of difficult to implement a function like
```cpp
template<typename NewScalar>
ActionModel<NewScalar> cast()
```
because we are using virtualization in the classes. Perhaps we could discuss this later @jcarpent